Adding normalized liquidity calculation for token pairs with BPT incl…

lgahdl opened the pull request:
…uded(Add/Remove Liquidity transactions);

Commit:
Adding normalized liquidity calculation for token pairs with BPT included(Add/Remove Liquidity transactions);

Commit:
running prettier on file;

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Adding pools as graph nodes;

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

brunoguerios commented on modules/sources/contracts/fetch-tokenpair-data.ts:
```diff
@@ -193,11 +217,58 @@ function generateTokenPairs(filteredPools: PoolInput[]): TokenPair[] {
+        for (const poolToken of pool.tokens) {
+            // create all pairs for pool's bpt
+            //we don't want a pair of the bpt with itself
+            if (poolToken.address === pool.address || poolToken.address === pool.address) continue;
```
```suggestion
            if (poolToken.address === pool.address) continue;
```

brunoguerios replied:
In fact you won't need this check because no v3 pool will contain its BPT.

brunoguerios commented on modules/sources/contracts/fetch-tokenpair-data.ts:
```diff
@@ -98,12 +102,27 @@ export async function fetchTokenPairData(
+            // tokenA->tokenB with 1% of tokenA balance
+            bptTokenPair.aToBAmountIn = parseUnits(bptTokenPair.tokenA.balance, bptTokenPair.tokenA.decimals) / 100n;
+            // tokenA->tokenB with 100USD worth of tokenA
+            const oneHundredUsdOfTokenA = (parseFloat(bptTokenPair.tokenA.balance) / bptTokenPair.tokenA.balanceUsd) * 100;
+            bptTokenPair.effectivePriceAmountIn = parseUnits(`${oneHundredUsdOfTokenA}`, bptTokenPair.tokenA.decimals);
```
One thing that called my attention is that you're using BPT as the tokenA, which is the one that you need to know the `balanceUsd`. For some reason I think this info is not as available/updated for BPTs compared to "regular" pool tokens.
We could reverse the order and have BPT being tokenB for this whole flow without much effort, right @lgahdl?
@franzns - do you see benefit on this change? Or is it ok to leave as is?

brunoguerios replied:
Never mind. I saw that you're using `totalLiquidity` as `balanceUsd` and I think this should be ok 👍

brunoguerios reviewed (commented):
Logic seems to make sense 👍 
I left a comment about an if statement that will always be false, so you should be able to remove it.
Other than that, I'll leave for the API team to make a final review/approval on their end.

lgahdl committed:
Update modules/sources/contracts/fetch-tokenpair-data.ts

Co-authored-by: Bruno <bruno@balancerlabs.dev>

Commit:
Building Paths considering Add/Remove Liquidity Operations;

Commit:
Updating Balancer SDK package to 0.14.0;
Adding filter to V3 pool types compatible with add/remove liquidity paths;
Adapting swap query output to V0.14.0 query output;

Commit:
Merge commit '4905b2c4e06a51a5bfc6eea2fd57b45c7234c6df' into v3-sor-add-remove-liq

Commit:
Merge commit 'acac63e1883a655cfca29d04b905c701883a9142' into v3-sor-add-remove-liq

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

Commit:
Removing statement that will always be false (V3 Pools will not have BPT);

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

Commit:
Merge branch 'v3-sor-with-swaps' into normalized-liquidity-bpts

Commit:
Merge branch 'normalized-liquidity-bpts' into v3-sor-add-remove-liq

Commit:
Removing log;

Commit:
Merge pull request #297 from balancer/v3-sor-add-remove-liq

V3 SOR - Adapting V3 SOR to build paths with Add/Remove Liquidity Operations;
